File: 15/part1.pl


$prev_a = 703;
$prev_b = 516;

$factor_a = 16807;
$factor_b = 48271;

$matches = 0;

for ($i = 0; $i < 40000000; $i++) {

	
	$prev_a = generateValue($prev_a, $factor_a, 2147483647);
	
	$prev_b = generateValue($prev_b, $factor_b, 2147483647);

	
	if ((substr(dec2bin($prev_a ^ $prev_b), -16) ^ 0) == 0 ) {
		$matches++;
	}
}

print $matches . "\n";

sub dec2bin {
	my $str = unpack("B32", pack("N", shift));
    return $str;
}

sub generateValue {

	my ($value, $factor, $divisor) = @_;

	return ($value * $factor) % $divisor
}

File: 15/part1.php
<?php

$prev_a = 65;
$prev_b = 8921;

$factor_a = 16807;
$factor_b = 48271;

$matches = 0;

for ($i = 0; $i < 1000000; $i++) {

	$prev_a = generateValue($prev_a, $factor_a, 2147483647);
	$a_bits = substr(decbin((int) $prev_a), -16);

	$prev_b = generateValue($prev_b, $factor_b, 2147483647);
	$b_bits = substr(decbin((int) $prev_b), -16);

	if ($a_bits === $b_bits) {
		$matches++;
	}
}

echo $matches . "\n";

function generateValue($value, $factor, $divisor) {
	return bcmod(bcmul($value, $factor), $divisor);
}